_posts/2015-03-23-design-patterns-part-2.markdown

```diff
@@ -0,0 +1,155 @@
+method on it named "welcome\_email").
+
+We might end up with a context like this.
+
+{% highlight ruby %}
+
+class CreateUser
+  def initialize(repo, mailer)
+    @repo = repo
+     @mailer = mailer
+  end
+
+  def call(first_name, last_name, email)
+    user = @repo.create(first_name, last_name, email)
+    @mailer.welcome_email(user.first_name, user.email)
+    user
+  end
+
+end
+
+{% endhighlight %}
+
+To use this context, we’d instantiate our context with the colloborator classes we need.
+
+{% highlight ruby %}
+
+context = CreateUser.new(ARRepository.new, UserMail.new)
+context.call(first_name, last_name, email)
+
+{% endhighlight %}
+
+There are two big benefits to using dependency injection here.  First, if we decide we like Sequel more than ActiveRecord, it'd be easy to just swap in a different repository and have the class behave exactly the same way.  The code below should work the same as the code above.
+
+{% highlight ruby %}
+
+context = CreateUser.new(SequelRepository.new, UserMail.new)
+context.call(first_name, last_name, email)
+
+{% endhighlight %}
+
+The second big benefit is testing.  In my tests for the class, I don't have to inject the class with any Rails objects, like ActionMailer.  I can just use a mock object.  In the code below, I’ve setup two mock objects: one for the repositiory and one for the mailer. 
+
+{% highlight ruby %}
+
+require  'create_user'
+
+described CreateUser do
+  let(:repo) { double(:repo) }
+  let(:mailer) { double(:mailer) }
+
+  subject { described_class.new(repo, mailer) }
+
+end
+
+{% endhighlight %}
+
+Our tests will verify that the appropriate methods get called on the repo and the mailer.  This give us very granular control over testing the business logic in our class.  Namely, we don’t have to worry about problems in the collaborator classes negatively effecting our tests.  If there’s a problem in those collaborator classes it will be either their unit tests or the integration tests that reveal it.  But what we want in the tests for the CreateUser class is laser focus on our own business logic. This may not be complex logic now, but as the app grows, it can become elaborate.
+
+The second benefit is that tests run fast.  When you compare testing a PORO to running a test that depends on ActiveRecord, you’re going to find that it takes forever to load all of ActiveRecord.  The tests for your PORO class will be done running before you’re even done requiring ActiveRecord in your Rails class.  It’ll take even longer if you actually want to have your tests create records in the database.  You’d have to initialize your app everytime you want to run your tests.  You’ll find that this ultimately disincentivizes you from maintaining a thorough tests suite. On the other hand, when your unit tests finish in seconds and you’re able to dive deep into the process of hammering out your business logic through tests, you’re going to enjoy the process much more.
```

Comment:
If you want to be a big troll you can link to DHH talking about how tests totally aren't slow with AR and say that we don't agree based on experience: http://david.heinemeierhansson.com/2014/slow-database-test-fallacy.html

:trollface: